### PR TITLE
New Files API and Cafe Request Handler

### LINKED ISSUE #568: Switch cafe service back to direct p2p
With the recent `go-libp2p-kad-dht` dialing fixes (https://github.com/libp2p/go-libp2p-kad-dht/pull/237), we can stop using the cafe service HTTP wrapper. Being able to add a cafe by peer ID is much nice than by URL and allows for a simpler cafe setup.

### LINKED ISSUE #573: Add ability to add 'arbitrary' IPFS CIDs to a Thread
**Is your feature request related to a problem? Please describe.**
Currently, Textile assumes that, if adding via an existing CID hash, that that hash has already been indexed locally. But this isn't a hard requirement. It should be possible, within `getFileConfig` to fetch IPFS data behind a CID directly, and use that as input to a mill similar to how the APIs `openFile` method returns 'raw' data.

**Describe the solution you'd like**
Allow arbitrary 'raw' IPFS data behind a CID hash to be used in the milling process via `getFileConfig`.

**Describe alternatives you've considered**
Alternatives involve creating complex Schemas to support arbitrarily complex Schema hierarchies. This might be a nice thing to have in the future, but the approach proposed above gets us there faster and should satisfy most existing use-cases.

**Additional context**
This was a feature 'indirectly requested' by a new user, and seemed like low-enough hanging fruit that it was worth pursuing.

### LINKED ISSUE #737: Write CafeHTTPRequest body to file
Background uploads are only supported for file-based data. This makes sense because the app may be terminated during an upload, and therefore all the app’s memory goes away. We need to write the request payloads to disk and `CafeHTTPRequest` should just have a property that is the path to the file.

### LINKED ISSUE #760: Write CafeRequest payload to disk
This is in the case of delegating a CafeRequest to the mobile native sdks because background uploads can only happen from files (not in memory data).

### LINKED ISSUE #761: Provide CafeRequest "failed" api
In the use case of mobile sdks, if for some reason we just can't successfully process a pin request, we need to notify the go layer that it failed.

### LINKED ISSUE #765: Add an additional group Id to cafe requests (sync group)
See comments here: https://github.com/textileio/go-textile/issues/586#issuecomment-491623283

### LINKED ISSUE #766: Split block pinning and publishing
Block pinning and thread publishing needs to be split into distinct steps. Once the associated "sync group" has been completely synced, the block can be "released" to the thread.

See here: https://github.com/textileio/go-textile/issues/586#issuecomment-491623283

### LINKED ISSUE #769: Panic cause shutdown during thread remove over rest API
2019/05/13 14:59:21 [Recovery] 2019/05/13 - 14:59:21 panic recovered:
DELETE /api/v0/threads/12D3KooWJNZke6ecTk7aGMKXCYNyNj8RU74326fJBFvw79zruRtZ HTTP/1.1
Host: 127.0.0.1:40600
Connection: close
Accept: */*
Accept-Encoding: gzip,deflate
Connection: close
Content-Length: 0
User-Agent: node-fetch/1.0 (+https://github.com/bitinn/node-fetch)
X-Textile-Args: 
X-Textile-Opts: 

runtime error: slice bounds out of range
/usr/local/go/src/runtime/panic.go:54 (0x402d8aa)
	panicslice: panic(sliceError)
/Users/andrewhill/go/pkg/mod/github.com/textileio/go-textile@v0.1.13-rc8/core/threads.go:345 (0x4e29448)
	(*Textile).RemoveThread: copy(t.loadedThreads[index:], t.loadedThreads[index+1:])
/Users/andrewhill/go/pkg/mod/github.com/textileio/go-textile@v0.1.13-rc8/core/api_threads.go:236 (0x4de8239)
	(*api).rmThreads: if _, err := a.node.RemoveThread(id); err != nil {
/Users/andrewhill/go/pkg/mod/github.com/gin-gonic/gin@v1.3.0/context.go:108 (0x4731192)
	(*Context).Next: c.handlers[c.index](c)
/Users/andrewhill/go/pkg/mod/github.com/gin-gonic/gin@v1.3.0/recovery.go:48 (0x4742719)
	RecoveryWithWriter.func1: c.Next()
/Users/andrewhill/go/pkg/mod/github.com/gin-gonic/gin@v1.3.0/context.go:108 (0x4731192)
	(*Context).Next: c.handlers[c.index](c)
/Users/andrewhill/go/pkg/mod/github.com/gin-gonic/gin@v1.3.0/logger.go:84 (0x4741ade)
	LoggerWithWriter.func1: c.Next()
/Users/andrewhill/go/pkg/mod/github.com/gin-gonic/gin@v1.3.0/context.go:108 (0x4731192)
	(*Context).Next: c.handlers[c.index](c)
/Users/andrewhill/go/pkg/mod/github.com/gin-gonic/gin@v1.3.0/gin.go:361 (0x4739bd6)
	(*Engine).handleHTTPRequest: c.Next()
/Users/andrewhill/go/pkg/mod/github.com/gin-gonic/gin@v1.3.0/gin.go:326 (0x4739413)
	(*Engine).ServeHTTP: engine.handleHTTPRequest(c)
/usr/local/go/src/net/http/server.go:2774 (0x4325807)
	serverHandler.ServeHTTP: handler.ServeHTTP(rw, req)
/usr/local/go/src/net/http/server.go:1878 (0x43213f0)
	(*conn).serve: serverHandler{c.server}.ServeHTTP(w, w.req)
/usr/local/go/src/runtime/asm_amd64.s:1337 (0x405bac0)
	goexit: BYTE	$0x90	// NOP

### LINKED ISSUE #793: textile cafe messages
**Describe the bug**
command `textile cafe messages` returns "cafe not found" - not taking second argument?

**To Reproduce**

Add a cafe
Run textile cafe messages
See error

**Expected behavior**
Response from API as described in cli

**Additional context**
0.2.2

### PR DESCRIPTION
**Pull request by sanderpick**
Fixes #737  
Fixes #765 
Fixes #766 
Fixes #760 
Fixes #761 
Fixes #769 
Fixes #573
Fixes #793
Fixes #568 

### Files
- Allow adding data to a thread from an arbitrary IPFS hash. If this hash corresponds to indexed encrypted data, that will be used. Otherwise, the mill input is catted from the network.

### Search
- Adds a standalone search endpoint to the cafe v1 HTTP API (this should eventually move to p2p)

### Cafes
- Moves most cafe communication to p2p
- More intelligent block sync grouping
- Splits pinning and posting
- Allows blocks to be created and left in a "pending" state during their and associated files sync before being "released" to the thread and peers
- Adds "historical" sync after a cafe registration
- Unpins files when the associated block is deleted / ignored
- Simplifies and reduces the number of calls to flush
- Adds mobile sync events
- Includes a mobile cafe sync test

### Events
- `WalletUpdate` -> `AccountUpdate`

### Threads v1.1
In order to handle above sync logic, block parents had to be pulled out of the block. I took this as an opportunity to move to a simple IPLD structure that paves the way for threads v2. A thread "node" has a link for parents (an index map like `"0" -> CID`, etc.) and a link for the encrypted block. The other advantage here is that blocks that are unreachable do not hold up the traversal. This is a trade-off with keeping the links obfuscated. 

- Blocks are wrapped w/ an IPLD node
- Inbox requests contain the entire node (otherwise the block release would have to somehow wait on another pin). The receiving cafe just pins this, making it more readily available for the recipient.
- More intelligent merging: we simply maintain multiple heads per thread that are joined when an outbound update is created.

### Mobile Cafes API
- Re-attempting p2p-based registration

### Mobile Files API
No more prepare, add hell.

- `AddData`
- `AddFiles` (can add multiple)
- `ShareFiles` (can add multiple)
- `SetAvatar`
- allow raw IPFS hashes as input to `AddFiles` (REST API needs updating, can do separately)

### Tests
- Adds comprehensive sync test to both core (desktop handler) and mobile

### CODE REVIEW

**Comment by sanderpick** [comment]
FYI, new mobile Files API is shaping up!

- `AddData`
- `AddFiles` (can add multiple)
- `ShareFiles` (can add multiple)

- `SetAvatar`

No more prepare, add hell.

**Comment by sanderpick** [comment]
This is getting close! I can open for review tomorrow. I just need to finish the tests to ensure everything is pinned from the sync process.

**Comment by sanderpick** [comment]
@asutula, here's sync status coming over the mobile event bus from a new mobile cafes test suite:

```
=== MOBILE FLUSH CALLED
+++ MOBILE EVENT: CAFE_SYNC_GROUP_UPDATE
>>> QmNcmpX64TTnfPtgr2gMda6Vz1z76UY6e7B1rwp3bofYJb
num. pending: 1
num. complete: 0
num. total: 1
size pending: 297
size complete: 0
size total: 297
<<<
+++ MOBILE EVENT: CAFE_SYNC_GROUP_UPDATE
>>> 1Ldon5UeyqoxY7DEKCd0cf5pld6
num. pending: 1
num. complete: 0
num. total: 1
size pending: 0
size complete: 0
size total: 0
<<<
+++ MOBILE EVENT: CAFE_SYNC_GROUP_UPDATE
>>> QmUqMskrkVM32Ket8YDipC3xmjDkF9bBZ1TWnL2Zu1VWKp
num. pending: 2
num. complete: 0
num. total: 22
size pending: 45766
size complete: 0
size total: 159510
<<<
+++ MOBILE EVENT: CAFE_SYNC_GROUP_UPDATE
>>> 1LdonKPXjsEgQ9Hv74AJPtw014q
num. pending: 1
num. complete: 0
num. total: 1
size pending: 0
size complete: 0
size total: 0
<<<
+++ MOBILE EVENT: CAFE_SYNC_GROUP_UPDATE
>>> 1LdonT9EOpafoXJOn5DBJqh4ZpD
num. pending: 1
num. complete: 0
num. total: 1
size pending: 0
size complete: 0
size total: 0
<<<
+++ MOBILE EVENT: CAFE_SYNC_GROUP_UPDATE
>>> QmfHpycGr3eR716Nm7QMd1VkcQ41qYuzsFuyrbE75a25KZ
num. pending: 5
num. complete: 0
num. total: 12
size pending: 502
size complete: 0
size total: 52357
<<<
+++ MOBILE EVENT: CAFE_SYNC_GROUP_UPDATE
>>> QmfHpycGr3eR716Nm7QMd1VkcQ41qYuzsFuyrbE75a25KZ
num. pending: 7
num. complete: 0
num. total: 12
size pending: 9623
size complete: 0
size total: 52357
<<<
+++ MOBILE EVENT: CAFE_SYNC_GROUP_UPDATE
>>> QmUqMskrkVM32Ket8YDipC3xmjDkF9bBZ1TWnL2Zu1VWKp
num. pending: 11
num. complete: 0
num. total: 22
size pending: 46768
size complete: 0
size total: 159510
<<<
+++ MOBILE EVENT: CAFE_SYNC_GROUP_UPDATE
>>> QmUqMskrkVM32Ket8YDipC3xmjDkF9bBZ1TWnL2Zu1VWKp
num. pending: 13
num. complete: 0
num. total: 22
size pending: 86604
size complete: 0
size total: 159510
<<<
+++ MOBILE EVENT: CAFE_SYNC_GROUP_UPDATE
>>> QmTP3WMB3Dv9dGQLiChVD91c4BJDfqG5bK1Cr1MY8mKfHY
num. pending: 5
num. complete: 0
num. total: 12
size pending: 504
size complete: 0
size total: 107502
<<<
+++ MOBILE EVENT: CAFE_SYNC_GROUP_UPDATE
>>> QmTP3WMB3Dv9dGQLiChVD91c4BJDfqG5bK1Cr1MY8mKfHY
num. pending: 6
num. complete: 0
num. total: 12
size pending: 1077
size complete: 0
size total: 107502
<<<
+++ MOBILE EVENT: CAFE_SYNC_GROUP_UPDATE
>>> QmadqsLgbFVceRvHJFMPApd1H2dDVxC3aEXWn9qk3ybeKN
num. pending: 1
num. complete: 0
num. total: 1
size pending: 347
size complete: 0
size total: 347
<<<
+++ MOBILE EVENT: CAFE_SYNC_GROUP_UPDATE
>>> QmUqMskrkVM32Ket8YDipC3xmjDkF9bBZ1TWnL2Zu1VWKp
num. pending: 15
num. complete: 0
num. total: 22
size pending: 101497
size complete: 0
size total: 159510
<<<
+++ MOBILE EVENT: CAFE_SYNC_GROUP_UPDATE
>>> 1Ldomhbp0WnAcmLLZnFNumz6Abp
num. pending: 1
num. complete: 0
num. total: 1
size pending: 314
size complete: 0
size total: 314
<<<
+++ MOBILE EVENT: CAFE_SYNC_GROUP_UPDATE
>>> 1LdomcqRrJnopk3jPhlza8BVxsY
num. pending: 1
num. complete: 0
num. total: 1
size pending: 0
size complete: 0
size total: 0
<<<
+++ MOBILE EVENT: CAFE_SYNC_GROUP_UPDATE
>>> QmfHpycGr3eR716Nm7QMd1VkcQ41qYuzsFuyrbE75a25KZ
num. pending: 8
num. complete: 0
num. total: 12
size pending: 10193
size complete: 0
size total: 52357
<<<
+++ MOBILE EVENT: CAFE_SYNC_GROUP_UPDATE
>>> 1Ldomz5Zc4dqNHTaqotsrDIXf5v
num. pending: 1
num. complete: 0
num. total: 1
size pending: 0
size complete: 0
size total: 0
<<<
+++ MOBILE EVENT: CAFE_SYNC_GROUP_UPDATE
>>> QmewUSCHJXLZ17fH1obLF6vqo63sDG4DpJ8bpLA1vxvCjX
num. pending: 1
num. complete: 0
num. total: 1
size pending: 338
size complete: 0
size total: 338
<<<
+++ MOBILE EVENT: CAFE_SYNC_GROUP_UPDATE
>>> 1LdonD6wqcNpDfQxWoZ5oYfL4YY
num. pending: 1
num. complete: 0
num. total: 1
size pending: 0
size complete: 0
size total: 0
<<<
+++ MOBILE EVENT: CAFE_SYNC_GROUP_UPDATE
>>> 1LdomdkJA3se3IqreLnFlzmeWsa
num. pending: 1
num. complete: 0
num. total: 1
size pending: 0
size complete: 0
size total: 0
<<<
+++ MOBILE EVENT: CAFE_SYNC_GROUP_UPDATE
>>> QmTP3WMB3Dv9dGQLiChVD91c4BJDfqG5bK1Cr1MY8mKfHY
num. pending: 8
num. complete: 0
num. total: 12
size pending: 46843
size complete: 0
size total: 107502
<<<
+++ MOBILE EVENT: CAFE_SYNC_GROUP_UPDATE
>>> QmTP3WMB3Dv9dGQLiChVD91c4BJDfqG5bK1Cr1MY8mKfHY
num. pending: 10
num. complete: 0
num. total: 12
size pending: 92609
size complete: 0
size total: 107502
<<<
+++ MOBILE EVENT: CAFE_SYNC_GROUP_UPDATE
>>> QmTP3WMB3Dv9dGQLiChVD91c4BJDfqG5bK1Cr1MY8mKfHY
num. pending: 12
num. complete: 0
num. total: 12
size pending: 107502
size complete: 0
size total: 107502
<<<
+++ MOBILE EVENT: CAFE_SYNC_GROUP_UPDATE
>>> QmaCSX9fP95PPYrC9Ym6roGxG8ZyTp1Yjd4yjMqVFsKD5f
num. pending: 1
num. complete: 0
num. total: 1
size pending: 594
size complete: 0
size total: 594
<<<
+++ MOBILE EVENT: CAFE_SYNC_GROUP_UPDATE
>>> QmUqMskrkVM32Ket8YDipC3xmjDkF9bBZ1TWnL2Zu1VWKp
num. pending: 17
num. complete: 0
num. total: 22
size pending: 103825
size complete: 0
size total: 159510
<<<
+++ MOBILE EVENT: CAFE_SYNC_GROUP_UPDATE
>>> QmUqMskrkVM32Ket8YDipC3xmjDkF9bBZ1TWnL2Zu1VWKp
num. pending: 19
num. complete: 0
num. total: 22
size pending: 112946
size complete: 0
size total: 159510
<<<
+++ MOBILE EVENT: CAFE_SYNC_GROUP_UPDATE
>>> Qmbi4ATd7jYmYxdkQ8fgtgnSA9r6z4Srr32bmU3xfw1Fi7
num. pending: 1
num. complete: 0
num. total: 1
size pending: 297
size complete: 0
size total: 297
<<<
+++ MOBILE EVENT: CAFE_SYNC_GROUP_UPDATE
>>> Qmbg1J3VGx8NXcyU2fdMqnrLpz8eqQ3yfMzL7S698nBrUc
num. pending: 1
num. complete: 0
num. total: 1
size pending: 596
size complete: 0
size total: 596
<<<
+++ MOBILE EVENT: CAFE_SYNC_GROUP_UPDATE
>>> QmfHpycGr3eR716Nm7QMd1VkcQ41qYuzsFuyrbE75a25KZ
num. pending: 10
num. complete: 0
num. total: 12
size pending: 50029
size complete: 0
size total: 52357
<<<
+++ MOBILE EVENT: CAFE_SYNC_GROUP_UPDATE
>>> QmfHpycGr3eR716Nm7QMd1VkcQ41qYuzsFuyrbE75a25KZ
num. pending: 12
num. complete: 0
num. total: 12
size pending: 52357
size complete: 0
size total: 52357
<<<
+++ MOBILE EVENT: CAFE_SYNC_GROUP_UPDATE
>>> 1LdonBwWKtLlZL5e6slCqYTTQGN
num. pending: 1
num. complete: 0
num. total: 1
size pending: 0
size complete: 0
size total: 0
<<<
+++ MOBILE EVENT: CAFE_SYNC_GROUP_UPDATE
>>> QmUqMskrkVM32Ket8YDipC3xmjDkF9bBZ1TWnL2Zu1VWKp
num. pending: 20
num. complete: 0
num. total: 22
size pending: 113744
size complete: 0
size total: 159510
<<<
+++ MOBILE EVENT: CAFE_SYNC_GROUP_UPDATE
>>> 1LdomdTjsXbIxQpGIgerIFO1Mn4
num. pending: 1
num. complete: 0
num. total: 1
size pending: 0
size complete: 0
size total: 0
<<<
+++ MOBILE EVENT: CAFE_SYNC_GROUP_UPDATE
>>> QmUqMskrkVM32Ket8YDipC3xmjDkF9bBZ1TWnL2Zu1VWKp
num. pending: 22
num. complete: 0
num. total: 22
size pending: 159510
size complete: 0
size total: 159510
<<<
+++ MOBILE EVENT: CAFE_SYNC_GROUP_UPDATE
>>> 1LdonRG11cmG0i6w5FmwoVmxv1i
num. pending: 1
num. complete: 0
num. total: 1
size pending: 0
size complete: 0
size total: 0
<<<
+++ MOBILE EVENT: CAFE_SYNC_GROUP_UPDATE
>>> QmUqMskrkVM32Ket8YDipC3xmjDkF9bBZ1TWnL2Zu1VWKp
num. pending: 21
num. complete: 1
num. total: 22
size pending: 158712
size complete: 798
size total: 159510
<<<
+++ MOBILE EVENT: CAFE_SYNC_GROUP_COMPLETE
>>> 1LdomdTjsXbIxQpGIgerIFO1Mn4
num. pending: 0
num. complete: 1
num. total: 1
size pending: 0
size complete: 0
size total: 0
<<<
+++ MOBILE EVENT: CAFE_SYNC_GROUP_UPDATE
>>> QmUqMskrkVM32Ket8YDipC3xmjDkF9bBZ1TWnL2Zu1VWKp
num. pending: 19
num. complete: 3
num. total: 22
size pending: 112946
size complete: 46564
size total: 159510
<<<
+++ MOBILE EVENT: CAFE_SYNC_GROUP_COMPLETE
>>> 1LdonRG11cmG0i6w5FmwoVmxv1i
num. pending: 0
num. complete: 1
num. total: 1
size pending: 0
size complete: 0
size total: 0
<<<
+++ MOBILE EVENT: CAFE_SYNC_GROUP_COMPLETE
>>> QmNcmpX64TTnfPtgr2gMda6Vz1z76UY6e7B1rwp3bofYJb
num. pending: 0
num. complete: 1
num. total: 1
size pending: 0
size complete: 297
size total: 297
<<<
+++ MOBILE EVENT: CAFE_SYNC_GROUP_COMPLETE
>>> 1Ldon5UeyqoxY7DEKCd0cf5pld6
num. pending: 0
num. complete: 1
num. total: 1
size pending: 0
size complete: 0
size total: 0
+++ MOBILE EVENT: CAFE_SYNC_GROUP_UPDATE
>>> QmUqMskrkVM32Ket8YDipC3xmjDkF9bBZ1TWnL2Zu1VWKp
num. pending: 17
num. complete: 5
num. total: 22
size pending: 67180
size complete: 92330
size total: 159510
<<<
+++ MOBILE EVENT: CAFE_SYNC_GROUP_COMPLETE
>>> 1LdonT9EOpafoXJOn5DBJqh4ZpD
num. pending: 0
num. complete: 1
num. total: 1
size pending: 0
size complete: 0
size total: 0
<<<
+++ MOBILE EVENT: CAFE_SYNC_GROUP_UPDATE
>>> QmfHpycGr3eR716Nm7QMd1VkcQ41qYuzsFuyrbE75a25KZ
num. pending: 7
num. complete: 5
num. total: 12
size pending: 51855
size complete: 502
size total: 52357
<<<
+++ MOBILE EVENT: CAFE_SYNC_GROUP_UPDATE
>>> QmfHpycGr3eR716Nm7QMd1VkcQ41qYuzsFuyrbE75a25KZ
num. pending: 5
num. complete: 7
num. total: 12
size pending: 42734
size complete: 9623
size total: 52357
<<<
+++ MOBILE EVENT: CAFE_SYNC_GROUP_UPDATE
>>> QmUqMskrkVM32Ket8YDipC3xmjDkF9bBZ1TWnL2Zu1VWKp
num. pending: 8
num. complete: 14
num. total: 22
size pending: 66178
size complete: 93332
size total: 159510
<<<
+++ MOBILE EVENT: CAFE_SYNC_GROUP_UPDATE
>>> QmUqMskrkVM32Ket8YDipC3xmjDkF9bBZ1TWnL2Zu1VWKp
num. pending: 6
num. complete: 16
num. total: 22
size pending: 26342
size complete: 133168
size total: 159510
<<<
+++ MOBILE EVENT: CAFE_SYNC_GROUP_COMPLETE
>>> 1LdonKPXjsEgQ9Hv74AJPtw014q
num. pending: 0
num. complete: 1
num. total: 1
size pending: 0
size complete: 0
size total: 0
<<<
+++ MOBILE EVENT: CAFE_SYNC_GROUP_UPDATE
>>> QmTP3WMB3Dv9dGQLiChVD91c4BJDfqG5bK1Cr1MY8mKfHY
num. pending: 11
num. complete: 1
num. total: 12
size pending: 106929
size complete: 573
size total: 107502
<<<
+++ MOBILE EVENT: CAFE_SYNC_GROUP_COMPLETE
>>> QmadqsLgbFVceRvHJFMPApd1H2dDVxC3aEXWn9qk3ybeKN
num. pending: 0
num. complete: 1
num. total: 1
size pending: 0
size complete: 347
size total: 347
<<<
+++ MOBILE EVENT: CAFE_SYNC_GROUP_UPDATE
>>> QmUqMskrkVM32Ket8YDipC3xmjDkF9bBZ1TWnL2Zu1VWKp
num. pending: 4
num. complete: 18
num. total: 22
size pending: 11449
size complete: 148061
size total: 159510
<<<
+++ MOBILE EVENT: CAFE_SYNC_GROUP_COMPLETE
>>> 1Ldomhbp0WnAcmLLZnFNumz6Abp
num. pending: 0
num. complete: 1
num. total: 1
size pending: 0
size complete: 314
size total: 314
<<<
+++ MOBILE EVENT: CAFE_SYNC_GROUP_COMPLETE
>>> 1LdomcqRrJnopk3jPhlza8BVxsY
num. pending: 0
num. complete: 1
num. total: 1
size pending: 0
size complete: 0
size total: 0
<<<
+++ MOBILE EVENT: CAFE_SYNC_GROUP_UPDATE
>>> QmfHpycGr3eR716Nm7QMd1VkcQ41qYuzsFuyrbE75a25KZ
num. pending: 4
num. complete: 8
num. total: 12
size pending: 42164
size complete: 10193
size total: 52357
<<<
+++ MOBILE EVENT: CAFE_SYNC_GROUP_COMPLETE
>>> 1Ldomz5Zc4dqNHTaqotsrDIXf5v
num. pending: 0
num. complete: 1
num. total: 1
size pending: 0
size complete: 0
size total: 0
<<<
+++ MOBILE EVENT: CAFE_SYNC_GROUP_UPDATE
>>> QmTP3WMB3Dv9dGQLiChVD91c4BJDfqG5bK1Cr1MY8mKfHY
num. pending: 6
num. complete: 6
num. total: 12
size pending: 106425
size complete: 1077
size total: 107502
<<<
+++ MOBILE EVENT: CAFE_SYNC_GROUP_COMPLETE
>>> QmewUSCHJXLZ17fH1obLF6vqo63sDG4DpJ8bpLA1vxvCjX
num. pending: 0
num. complete: 1
num. total: 1
size pending: 0
size complete: 338
size total: 338
<<<
+++ MOBILE EVENT: CAFE_SYNC_GROUP_COMPLETE
>>> 1LdomdkJA3se3IqreLnFlzmeWsa
num. pending: 0
num. complete: 1
num. total: 1
size pending: 0
size complete: 0
size total: 0
<<<
+++ MOBILE EVENT: CAFE_SYNC_GROUP_UPDATE
>>> QmTP3WMB3Dv9dGQLiChVD91c4BJDfqG5bK1Cr1MY8mKfHY
num. pending: 4
num. complete: 8
num. total: 12
size pending: 60659
size complete: 46843
size total: 107502
<<<
+++ MOBILE EVENT: CAFE_SYNC_GROUP_UPDATE
>>> QmTP3WMB3Dv9dGQLiChVD91c4BJDfqG5bK1Cr1MY8mKfHY
num. pending: 2
num. complete: 10
num. total: 12
size pending: 14893
size complete: 92609
size total: 107502
<<<
+++ MOBILE EVENT: CAFE_SYNC_GROUP_COMPLETE
>>> QmTP3WMB3Dv9dGQLiChVD91c4BJDfqG5bK1Cr1MY8mKfHY
num. pending: 0
num. complete: 12
num. total: 12
size pending: 0
size complete: 107502
size total: 107502
<<<
+++ MOBILE EVENT: CAFE_SYNC_GROUP_COMPLETE
>>> 1LdonD6wqcNpDfQxWoZ5oYfL4YY
num. pending: 0
num. complete: 1
num. total: 1
size pending: 0
size complete: 0
size total: 0
<<<
+++ MOBILE EVENT: CAFE_SYNC_GROUP_COMPLETE
>>> QmaCSX9fP95PPYrC9Ym6roGxG8ZyTp1Yjd4yjMqVFsKD5f
num. pending: 0
num. complete: 1
num. total: 1
size pending: 0
size complete: 594
size total: 594
<<<
+++ MOBILE EVENT: CAFE_SYNC_GROUP_UPDATE
>>> QmUqMskrkVM32Ket8YDipC3xmjDkF9bBZ1TWnL2Zu1VWKp
num. pending: 2
num. complete: 20
num. total: 22
size pending: 9121
size complete: 150389
size total: 159510
<<<
+++ MOBILE EVENT: CAFE_SYNC_GROUP_COMPLETE
>>> Qmbi4ATd7jYmYxdkQ8fgtgnSA9r6z4Srr32bmU3xfw1Fi7
num. pending: 0
num. complete: 1
num. total: 1
size pending: 0
size complete: 297
size total: 297
<<<
+++ MOBILE EVENT: CAFE_SYNC_GROUP_COMPLETE
>>> Qmbg1J3VGx8NXcyU2fdMqnrLpz8eqQ3yfMzL7S698nBrUc
num. pending: 0
num. complete: 1
num. total: 1
size pending: 0
size complete: 596
size total: 596
<<<
+++ MOBILE EVENT: CAFE_SYNC_GROUP_UPDATE
>>> QmfHpycGr3eR716Nm7QMd1VkcQ41qYuzsFuyrbE75a25KZ
num. pending: 2
num. complete: 10
num. total: 12
size pending: 2328
size complete: 50029
size total: 52357
<<<
+++ MOBILE EVENT: CAFE_SYNC_GROUP_COMPLETE
>>> QmfHpycGr3eR716Nm7QMd1VkcQ41qYuzsFuyrbE75a25KZ
num. pending: 0
num. complete: 12
num. total: 12
size pending: 0
size complete: 52357
size total: 52357
<<<
+++ MOBILE EVENT: CAFE_SYNC_GROUP_COMPLETE
>>> 1LdonBwWKtLlZL5e6slCqYTTQGN
num. pending: 0
num. complete: 1
num. total: 1
size pending: 0
size complete: 0
size total: 0
<<<
+++ MOBILE EVENT: CAFE_SYNC_GROUP_COMPLETE
>>> QmUqMskrkVM32Ket8YDipC3xmjDkF9bBZ1TWnL2Zu1VWKp
num. pending: 0
num. complete: 22
num. total: 22
size pending: 0
size complete: 159510
size total: 159510
<<<
```

**Comment by asutula** [comment]
Looking awesome. In cases where num pending + num complete != num total, does that mean the unaccounted-for ones are being processed externally that that moment (by the native sdk, for example)?

**Comment by sanderpick** [comment]
The group requests are limited by some `int`. So, that case you call out just means that there were extra not returned in the group (they were not set to pending).

File: mobile/cafes_test.go [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+
+/*
+Handle the request queue.
+  1. List some groups
+  2. List get the HTTP request list for each of those groups
+  3. Handle them (set to pending, send to cafe)
+  4. Delete failed (reties not handled here)
+  5. Set successful to complete
+*/
+func flushCafeRequest(limit int) (int, error) {
```
Comment (sanderpick) [inline mobile/cafes_test.go]: ^^ @asutula here's the process to implement in the SDK.  (fyi, this test method should read `flushCafeRequests` (plural)). The idea is that you would list the groups until there are zero.

Comment (sanderpick) [reply mobile/cafes_test.go]: I'm going to remove that "group" terminology in the tests here cause it's not needed to understand the process.

File: mobile/cafes_test.go [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+}
+
+/*
+Handle the request queue.
+  1. List some requests
+  2. Write the HTTP request for each
+  3. Handle them (set to pending, send to cafe)
+  4. Delete failed (reties not handled here)
+  5. Set successful to complete
+*/
```
Comment (sanderpick) [inline mobile/cafes_test.go]: Ok, this is the process sans typos.

File: mobile/cafes_test.go [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+		}
+		if err != nil {
+			return count, err
+		}
+		res.Body.Close()
+	}
+	return count, nil
+}
+
+func flush(batchSize int) error {
```
Comment (sanderpick) [inline mobile/cafes_test.go]: This test method here just manually recursively handles all the requests after adding a bunch of data.

**Comment by sanderpick** [comment]
TODO:

- [ ] Finish desktop requests handler
- [ ] Deprecate the `/service` cafe HTTP endpoint
- [ ] Desktop -> p2p where possible
- [ ] Add file download queue
- [ ] Add a datastore migration

File: core/thread.go [hunk 4]
```diff
@@ -543,42 +538,35 @@ func (t *Thread) indexBlock(commit *commitResult, blockType pb.Block_BlockType,
...
-// handleHead determines whether or not a thread can be fast-forwarded or if a merge block is needed
-// - parents are the parents of the incoming chain
-func (t *Thread) handleHead(inbound mh.Multihash, parents []string) (mh.Multihash, error) {
-	head, err := t.Head()
+// handleHead determines what the next set of HEADs will be.
+// One of three situations will occur:
+// 1) fast-forward: the inbound parents are identical to current heads (heads -> [inbound])
+// 2) partial-split: at least one inbound parent matches a current head (heads -> [inbound, unmatched...])
+// 3) full-split: zero inbound parents match a current head (heads -> [inbound, current...])
+func (t *Thread) handleHead(inbound string, inboundParents []string) error {
```
Comment (sanderpick) [inline core/thread.go]: @carsonfarmer will need your eyes to double check me here. This feels much better... no more merge blocks, just maintain multiple HEADs until they can be tied back together with a local update (or a magical remote update (situation 1).

Comment (carsonfarmer) [reply core/thread.go]: I also noticed it is possible to have a 'pending' parent. Is this just the default until a remote or local update comes in?

**Comment by sanderpick** [comment]
Taking this out of WIP. I'll add a bit more the changelog after a call.

**Comment by sanderpick** [comment]
Ok, I've updated the change log and bumped the version to `0.3.0`. The block download queue work is living on a different branch (https://github.com/textileio/go-textile/tree/sander/block-queue). It demanded more threads changes and I didn't want to further balloon this PR.

File: core/blocks.go [hunk 5]
```diff
@@ -32,6 +32,19 @@ func (t *Textile) Block(id string) (*pb.Block, error) {
 	return block, nil
 }
 
+// Block returns block with id
```
Comment (carsonfarmer) [inline core/blocks.go]: Nice, been digging into this already, very useful!

File: cmd/main.go [hunk 6]
```diff
@@ -127,8 +127,8 @@ var (
 	// add
 	cafeAddCmd = cafeCmd.Command("add", `Registers with a cafe and saves an expiring service session token.
 An access token is required to register, and should be obtained separately from the target cafe.`)
-	cafeAddURL   = cafeAddCmd.Arg("url", "The host cafe public URL").Required().String()
-	cafeAddToken = cafeAddCmd.Flag("token", "An access token supplied by the cafe").Short('t').Required().String()
+	cafeAddPeerID = cafeAddCmd.Arg("peer", "The host cafe's IPFS peer ID").Required().String()
```
Comment (carsonfarmer) [inline cmd/main.go]: Nice to have this back!

File: core/cafe_api_v1.go [hunk 7]
```diff
@@ -97,49 +110,53 @@ func (c *cafeApi) storeThread(g *gin.Context) {
...
 
+	log.Debugf("unstored thread %s", id)
+
 	g.Status(http.StatusNoContent)
 }
 
 func (c *cafeApi) deliverMessage(g *gin.Context) {
-	pid := g.Request.Header.Get("X-Textile-Peer")
-	clientId := g.Param("pid")
+	from := g.Param("from")
```
Comment (carsonfarmer) [inline core/cafe_api_v1.go]: Nice!

**Comment by carsonfarmer** [comment]
Minor version.go conflict

**Comment by balupton** [comment]
When is this aimed to land?

I may cherry-pick https://github.com/textileio/go-textile/issues/793 in #812 if there is a chance 812 will land sooner.

**Comment by sanderpick** [comment]
Finally was able to rebase this... can merge once tests pass. Can I get another approval?

**Pull request merged**